### ISSUE TITLE
Remaining-characters counter on input field

### ISSUE THREAD

**Issue by dgw** [thelounge/thelounge#3331]
I'm posting this partially on behalf of a few other users in one of my IRC channels who agreed that this would be a good idea, after I idly brought it up immediately following a one-word-too-long message that The Lounge split into two lines.

### Feature Description
As the input text approaches the maximum length of a single message, the input field starts to display a small remaining-character count and the number of messages that will be sent.

I'm inspired by the behavior of mobile phone SMS clients, many of which I've observed start showing a counter at  "20" when a message reaches 140 characters, and count down from there to "3", "2", "1", "0", "157&nbsp;/&nbsp;2", and so forth.<sup>1</sup>

It could be helpful for The Lounge to show the user when they've typed something approaching the maximum length of an IRC message—as determined by The Lounge's line-splitting logic, taking word breaks into account. This means it would be slightly more complicated than "just" counting characters and taking a modulo.

As I understand it, SMS clients display this mostly to keep informed users who still pay for SMS by-the-message (each part of a split message counts against their quota, and the recipient's too).

While The Lounge & IRC have the interplay of long, line-split messages & flood controls to consider, I envision this first as a writing aid. That said, when pasting a large piece of text (a quote from some website, say), it could also be useful to see how many messages would be sent after splitting it up.

Maybe The Lounge could adopt what Google Voice used to do, and after exceeding some number of pieces change the message count displayed to "Pastebin it!" (Unfortunately, the modern Google Voice interface doesn't do this character-count stuff any more. But the old version would change from showing the count to "Really?" if the user entered a message longer than about 7 parts.)

----

1. Protocol overhead eats up a few bytes of each part in a multi-part SMS message. SMS, of course, splits messages arbitrarily anywhere, and the parts get reconstructed by the receiving device into a single message. The Lounge has to be smart about word boundaries.

**Comment by MiniDigger** [comment]
> Maybe The Lounge could adopt what Google Voice used to do, and after exceeding some number of pieces change the message count displayed to "Pastebin it!"
and then not use a 3rd party bin but just use the same system as for image uploads

**Comment by xPaw** [comment]
> as determined by The Lounge's line-splitting logic, taking word breaks into account. This means it would be slightly more complicated than "just" counting characters and taking a modulo.

It would be quite complicated, because currently it takes into account not just word boundaries, but also graphemes, and there's a few other metrics. Taking all these into account would yield a really inconsistent character counter in my opinion (it would probably jump around constantly as it recalculcated).

> That said, when pasting a large piece of text (a quote from some website, say), it could also be useful to see how many messages would be sent after splitting it up.

> Maybe The Lounge could adopt what Google Voice used to do, and after exceeding some number of pieces change the message count displayed to "Pastebin it!"

We did think about this, and as file uploading exists now, it's certainly doable to suggest uploading long texts as a file, and you don't need the counter for it (just check full string length, or how many new lines there are).

**Comment by dgw** [comment]
Ran into the need for this a few times this week, so came back to check on it.

> Not to mention it would be rather expensive to run on every key press on the client

How about setting a brief timeout instead, and only running the expensive update when the user has paused typing? That's probably more of a UX question, then. (Does the counter continue to display the old value until typing pauses long enough to update, or clear itself on the first keypress and show some indeterminate state until the user stops typing? Etc.)

I'd also like to point out that Twitter has almost certainly implemented (or found in a library) some reasonably efficient method of doing something like this. Pasting the emoji "😜" 140 times into the tweet box (which allows 280 "characters" in theory) got me a "0" in their remainder display. I pasted the first 20 individually, then copied those and pasted 6 more batches of 20, to get there, and at no point did the counter fall behind.

All that said, I understand if no one else really cares about this feature, and will probably (finally) have time to look into it myself some time early next year.
